File: apps/api/plane/db/models/project_update.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

# Django imports
from django.db import models

# Module imports
from .project import ProjectBaseModel


class AIGenerationStatus(models.TextChoices):
    """Shared status choices for a single AI-drafted-update generation
    attempt - used both on `ProjectUpdate.ai_generation_status` (the latest
    attempt behind that update's current draft) and on `AIGenerationLog`
    (one row per attempt, see `plane.db.models.ai_generation_log`).
    """

    PENDING = "PENDING", "Pending"
    SUCCESS = "SUCCESS", "Success"
    FAILED = "FAILED", "Failed"
    TIMEOUT = "TIMEOUT", "Timeout"


class ProjectUpdate(ProjectBaseModel):
    """
    A historized, manually-authored status update for a project (On
    Track/At Risk/Off Track + free text) - see
    docs/feature-specs/03-projects-roadmaps-initiatives.md ("Mises a jour
    de statut structurees") in plane-selfhost. Deliberately a plain
    description_html (Initiative/Milestone's tier), not the collaborative
    quad-field pattern (description_json/_binary/_stripped) used by
    Issue/Page - a project update is authored by a single person at a
    point in time, not co-edited live, so wiring the apps/live Y.js
    service for this entity would be disproportionate.
    """

    class StatusChoice(models.TextChoices):
        ON_TRACK = "ON_TRACK", "On Track"
        AT_RISK = "AT_RISK", "At Risk"
        OFF_TRACK = "OFF_TRACK", "Off Track"

    status = models.CharField(max_length=20, choices=StatusChoice.choices)
    description_html = models.TextField(blank=True, default="<p></p>")
    # Snapshot of the auto-generated "since last update" block at creation
    # time - frozen, never regenerated on edit (exigence 6 de la spec).
    generated_summary_json = models.JSONField(default=dict, blank=True)
    is_summary_edited = models.BooleanField(default=False)

    # AI-assisted drafting fields - category 9 feature 6 ("Redaction
    # assistee des mises a jour de statut", docs/feature-specs/09-ai-features.md
    # in plane-selfhost). PROJECT ONLY in v1 - see
    # plane.utils.project_update_ai_draft module docstring for why
    # Cycle/Module are a deliberate, documented gap rather than an
    # oversight. Publication is unchanged (this endpoint's `create()` still
    # IS publication) - these fields are simply carried through from a
    # generated draft when the human submits the existing manual form.
    is_ai_assisted = models.BooleanField(default=False)
    # Raw generated text, kept verbatim even after the human edits the
    # published description_html - exigence 8's audit/diff requirement.
    ai_draft_content = models.TextField(null=True, blank=True)
    ai_generation_status = models.CharField(
        max_length=20, choices=AIGenerationStatus.choices, null=True, blank=True
    )
    # {llm_provider, llm_model, prompt_tokens, completion_tokens, latency_ms}
    # - see plane.utils.project_update_ai_draft.generate_project_update_draft
    # for the documented prompt_tokens/completion_tokens gap (the shared LLM
    # helper doesn't currently surface provider token usage).
    ai_generation_metadata = models.JSONField(null=True, blank=True)
    # Serialized issue-delta list actually used for this generation -
    # reproducibility/audit (exigence 8).
    ai_source_snapshot = models.JSONField(null=True, blank=True)
    ai_regeneration_count = models.PositiveSmallIntegerField(default=0)

    class Meta:
        verbose_name = "Project Update"
        verbose_name_plural = "Project Updates"
        db_table = "project_updates"
        ordering = ("-created_at",)

    def __str__(self):
        return f"{self.status} <{self.project_id}>"


class ProjectUpdateReminder(ProjectBaseModel):
    """
    Idempotency log for the periodic reminder task - one row per due
    date, so a periodic Celery beat task can safely run every few minutes
    without double-sending reminders for the same deadline.
    """

    scheduled_for = models.DateTimeField()
    sent_at = models.DateTimeField(null=True, blank=True)
    reminder_count = models.PositiveSmallIntegerField(default=0)

    class Meta:
        verbose_name = "Project Update Reminder"
        verbose_name_plural = "Project Update Reminders"
        db_table = "project_update_reminders"
        ordering = ("-created_at",)

    def __str__(self):
        return f"{self.project_id} <{self.scheduled_for}>"

File: packages/types/src/project-update/project-update.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

export type TProjectUpdateStatus = "ON_TRACK" | "AT_RISK" | "OFF_TRACK";

export type TProjectUpdateGeneratedSummary = {
  since: string;
  issues_created: number;
  issues_completed: number;
  issues_cancelled: number;
  net_backlog_change: number;
  cycles_started: number;
  cycles_closed: number;
};

/**
 * Category 9, feature 6 - "AI-assisted status update drafting"
 * (docs/feature-specs/09-ai-features.md in plane-selfhost, PROJECT-only in
 * this fork). Shared status for a single generation attempt - see
 * `AIGenerationStatus` (apps/api/plane/db/models/project_update.py).
 */
export type TAIGenerationStatus = "PENDING" | "SUCCESS" | "FAILED" | "TIMEOUT";

/** `ProjectUpdate.ai_generation_metadata` / `AIGenerationLog.token_usage`
 * shape - see `generate_project_update_draft`
 * (apps/api/plane/utils/project_update_ai_draft.py). `prompt_tokens`/
 * `completion_tokens` are always `null` today - documented gap, the shared
 * LLM helper doesn't currently surface provider token-usage counts. */
export type TProjectUpdateAIGenerationMetadata = {
  llm_provider?: string | null;
  llm_model?: string | null;
  prompt_tokens?: number | null;
  completion_tokens?: number | null;
  latency_ms?: number;
};

/** `ProjectUpdate.ai_source_snapshot` shape - the serialized issue-delta
 * actually used for a generation, frozen for reproducibility/audit - see
 * `compute_project_update_delta` (apps/api/plane/utils/project_update_ai_draft.py). */
export type TProjectUpdateAISourceSnapshot = {
  since: string;
  computed_at: string;
  data_scope: string;
  completed_issues: Record<string, unknown>[];
  overdue_issues: Record<string, unknown>[];
  blocked_issues: Record<string, unknown>[];
  remaining_distribution: Record<string, number>;
};

export interface IProjectUpdate {
  id: string;
  workspace_id: string;
  project_id: string;
  status: TProjectUpdateStatus;
  description_html: string;
  generated_summary_json: TProjectUpdateGeneratedSummary | Record<string, never>;
  is_summary_edited: boolean;
  // AI-assisted drafting provenance (exigence 8) - see `ProjectUpdate`'s own
  // field comments (apps/api/plane/db/models/project_update.py).
  is_ai_assisted: boolean;
  ai_draft_content: string | null;
  ai_generation_status: TAIGenerationStatus | null;
  ai_generation_metadata: TProjectUpdateAIGenerationMetadata | Record<string, never> | null;
  ai_source_snapshot: TProjectUpdateAISourceSnapshot | Record<string, never> | null;
  ai_regeneration_count: number;
  created_at: string;
  updated_at: string;
  created_by?: string;
  created_by_detail?: {
    id: string;
    first_name: string;
    last_name: string;
    display_name: string;
    avatar_url: string;
  };
}

export type TProjectUpdateWritePayload = {
  status: TProjectUpdateStatus;
  description_html: string;
  generated_summary_json?: TProjectUpdateGeneratedSummary | Record<string, never>;
  is_summary_edited?: boolean;
  // Only ever sent when publishing a draft generated via
  // `POST .../updates/draft/` - a manual, non-AI "Add update" submission
  // omits these entirely rather than forcing `is_ai_assisted: false`.
  is_ai_assisted?: boolean;
  ai_draft_content?: string | null;
  ai_generation_status?: TAIGenerationStatus | null;
  ai_generation_metadata?: TProjectUpdateAIGenerationMetadata | Record<string, never> | null;
  ai_source_snapshot?: TProjectUpdateAISourceSnapshot | Record<string, never> | null;
  ai_regeneration_count?: number;
};

export type TProjectUpdateCadence = "DISABLED" | "WEEKLY" | "BIWEEKLY" | "MONTHLY";

/** Response of `POST /api/workspaces/:slug/projects/:projectId/updates/draft/`
 * - never persists a `ProjectUpdate` row itself, see
 * `ProjectUpdateAIDraftEndpoint` (apps/api/plane/app/views/project_update/ai_draft.py).
 * `suggested_status` is a modifiable suggestion, never locked in. */
export type TProjectUpdateAIDraftResponse = {
  draft_content: string;
  suggested_status: TProjectUpdateStatus;
  ai_generation_status: TAIGenerationStatus;
  ai_generation_metadata: TProjectUpdateAIGenerationMetadata | Record<string, never>;
  ai_source_snapshot: TProjectUpdateAISourceSnapshot | Record<string, never>;
  regeneration_count: number;
  max_regenerations: number;
};

/** One row of `GET .../ai-update-logs/` (Admin-only, PROJECT level - same
 * gating as manual update creation, not a workspace-level permission) - see
 * `AIGenerationLogSerializer` (apps/api/plane/app/serializers/ai_generation_log.py). */
export type TAIGenerationLog = {
  id: string;
  workspace_id: string;
  project_id: string;
  update_id: string | null;
  status: TAIGenerationStatus;
  provider: string | null;
  model_name: string | null;
  token_usage: Record<string, unknown> | null;
  error_message: string | null;
  triggered_by: string | null;
  triggered_by_detail?: {
    id: string;
    first_name: string;
    last_name: string;
    display_name: string;
    avatar_url: string;
  };
  created_at: string;
};

export type TAIGenerationLogParams = {
  cursor?: string;
  per_page?: number;
};
